src/index.js

```diff
@@ -20,15 +17,10 @@ export function asString(raw) {
   return raw
 }
 
-export function asUndefined(raw) {
-  if (raw !== undefined) throw new TypeError('Unexpected value')
-  return undefined
-}
+export const asUndefined = asValue(undefined)
 
-export function asNone(raw) {
-  if (raw != null) throw new TypeError('Unexpected value')
-  return undefined
-}
+export const asNone = raw =>
+  asMaybe(asValue(null))(raw) === null ? undefined : asValue(undefined)(raw)
```

Comment:
This one can be `asOptional(asNull)`, which should do the exact same thing. Otherwise, the original version might have been easier to read.